torch/csrc/jit/passes/onnx/peephole.cpp

```diff
@@ -830,6 +830,113 @@ void removeMaxPoolUnusedOutput(Block* b) {
+    for (auto* child_block : it->blocks()) {
+      fuseLogSoftmaxNllLoss(child_block);
+    }
+    if (it->kind() == onnx::NegativeLogLikelihoodLoss) {
+      auto prev = it->input(0)->node();
+      Node* origNllLossNode = *it;
+      Node* origLogSoftmaxNode;
+      // if the input is 2D
+      // graph(%input : Float(3, 5),
+      // %target : Long(3)):
+      // %4 : Float(3, 5) = onnx::LogSoftmaxaxis=1
+      // %8 : Float(3) = onnx::NegativeLogLikelihoodLoss[reduction="none"]
+      // return (%8)
+      if (prev->kind() == onnx::LogSoftmax) {
+        origLogSoftmaxNode = it->input(0)->node();
+        // if the input is 4D
+        // graph(%input : Float(3, 5, 2, 7),
+        // %target : Long(3, 2, 7)):
+        // %4 : Tensor = onnx::Transpose[perm=[0, 3, 2, 1]] (%input)
+        // %5 : Tensor = onnx::LogSoftmax[axis=3] (%4)
+        // %6 : Float(3, 5, 2, 7) = onnx::Transpose[perm=[0, 3, 2, 1]] (%5)
+        // %10 : Float(3, 2, 7) =
+        // onnx::NegativeLogLikelihoodLoss[reduction="none"](%6, %target) return
+        // (%10)
+      } else if (
+          prev->kind() == onnx::Transpose &&
+          prev->input(0)->node()->kind() == onnx::LogSoftmax) {
+        origLogSoftmaxNode = prev->input(0)->node();
+        auto logSoftmaxNode = origLogSoftmaxNode->input(0)->node();
+        origLogSoftmaxNode->replaceInput(0, logSoftmaxNode->inputs().at(0));
+        // if the input is 3D or > 4D
+        // graph(%input : Float(3, 5, 2),
+        // %target.1 : Long(3, 2)):
+        // %4 : Tensor = onnx::Transpose[perm=[0, 2, 1]] (%input)
+        // %5 : Tensor = onnx::LogSoftmax[axis=2] (%4)
+        // %6 : Float(3, 5, 2) = onnx::Transpose[perm=[0, 2, 1]] (%5)
+        // ...
+        // %22 : Float(3, 5, 1, 2) = onnx::Reshape(%6, %21)
+        // ...
+        // %26 : Long(3, 1, 2) = onnx::Reshape(%target.1, %25)
+        // %30 : Float() = onnx::NegativeLogLikelihoodLoss[reduction="sum"](%22,
+        // %26) return (%30)
+      } else if (
+          prev->kind() == onnx::Reshape &&
+          prev->input(0)->node()->kind() == onnx::Transpose &&
+          prev->input(0)->node()->input(0)->node()->kind() ==
+              onnx::LogSoftmax) {
+        origLogSoftmaxNode = prev->input(0)->node()->input(0)->node();
+        auto logSoftmaxNode = origLogSoftmaxNode->input(0)->node();
+        origLogSoftmaxNode->replaceInput(0, logSoftmaxNode->inputs().at(0));
+        auto reshape = origNllLossNode->input(1)->node();
+        origNllLossNode->replaceInput(1, reshape->inputs().at(0));
+        // when reduction=none a different graph is created and the graph
+        // doesn't end with node NegativeLogLikelihoodLoss like in all other
+        // cases. graph(%input : Float(3, 5, 2), %target.1 : Long(3, 2)): %4 :
+        // Tensor = onnx::Transposeperm=[0, 2, 1] %5 : Tensor =
+        // onnx::LogSoftmaxaxis=2 %6 : Float(3, 5, 2) = onnx::Transposeperm=[0,
```

Comment:
nit: this part of comment seems compressed, is this intended?

Reply:
no, I will fix it